suite/tests/linux/rseq.c

```diff
@@ -130,18 +155,120 @@ test_rseq(void)
         /* clang-format off */ /* (avoid indenting next few lines) */
         ".long " STRINGIFY(RSEQ_SIG) "\n\t"
         "4:\n\t"
+        "addl $1, %[restarts]\n\t"
+        "movb $0, %[force_restart_write]\n\t"
         "jmp 6b\n\t"
 
         /* Clear the ptr. */
         "5:\n\t"
-        "movq $0, %0\n\t"
+        "movq $0, %[rseq_tls]\n\t"
         /* clang-format on */
 
-        : "=m"(rseq_tls.rseq_cs), "=m"(id), "=m"(restarts)
-        : "m"(rseq_tls.cpu_id), "i"(RSEQ_CPU_ID_UNINITIALIZED)
+        : [rseq_tls] "=m"(rseq_tls.rseq_cs), [id] "=m"(id),
+          [completions] "=m"(completions), [restarts] "=m"(restarts),
+          [force_restart_write] "=m"(force_restart)
+        : [cpu_id] "m"(rseq_tls.cpu_id), [cpu_id_uninit] "i"(RSEQ_CPU_ID_UNINITIALIZED),
+          [force_restart] "m"(force_restart)
         : "rax", "memory");
     assert(id != RSEQ_CPU_ID_UNINITIALIZED);
-    return restarts;
+    *completions_out = completions;
+    *restarts_out = restarts;
+}
+
+static void
+test_rseq_call(void)
+{
+    int completions, restarts;
+    sigill_count = 0;
+    test_rseq_call_once(false, &completions, &restarts);
+    /* There *could* have been a migration restart. */
+    assert(completions == 1 && sigill_count == 0);
+    test_rseq_call_once(true, &completions, &restarts);
+    assert(completions == 1 && restarts > 0 && sigill_count == 1);
+}
+
+static void
+test_rseq_branches_once(bool force_restart, int *completions_out, int *restarts_out)
+{
+    /* We use static to avoid stack reference issues with our extra frame inside the asm.
+     */
+    __u32 id = RSEQ_CPU_ID_UNINITIALIZED;
+    int completions = 0;
+    int restarts = 0;
+    __asm__ __volatile__(
+        /* clang-format off */ /* (avoid indenting next few lines) */
+        RSEQ_ADD_TABLE_ENTRY(branches, 2f, 3f, 4f)
```

Comment:
It is clang-format weirdness.  The comment after the RSEQ_ADD_TABLE_ENTRY at line 121 seems to prevent clang-format from doing anything unwanted.  For the other two instances, however, clang-format adds unwanted indentation.  Here is what happens for the one at line 200:

```C
        RSEQ_ADD_TABLE_ENTRY(branches, 2f, 3f, 4f)

            "6:\n\t"
            /* Store the entry into the ptr. */
            "leaq rseq_cs_branches(%%rip), %%rax\n\t"
```

These long string literal sequences with macros embedded are likely just a not-well-tested corner case to clang-format.

Are you looking for more comments to explain?  I think the existing comment about indenting the next few lines is reasonably sufficient.  Un-resolve if you disagree.